### PR TITLE
Callback plugin based on journald

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

This work is based upon the work of chestm007
Migrated from https://github.com/ansible/ansible/pull/43625

Callback plugin based on journald using systemd-python library.

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/journald_callback_plugin.yml
plugins/callback/journald.py

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @chestm007 I migrated journald callback plugin here.

@felixfontein @Andersson007 @gundalow Could you please review this ? Thanks in advance.

File: changelogs/fragments/journald_callback_plugin.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- journald - Add callback plugin based upon systemd-python.
```
Comment (felixfontein) [inline changelogs/fragments/journald_callback_plugin.yml:2]: This is not needed.

File: plugins/callback/journald.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    type: notification
+    short_description: Sends play events to systemd's journal
+    description:
+      - This is an ansible callback plugin that sends status updates to the system journal during playbook execution.
+    requirements:
+      - Whitelist in configuration
+      - systemd-python (Python library)
+    options:
+      logger_name:
+        required: False
```
Comment (Andersson007) [inline plugins/callback/journald.py:19]:
```suggestion
```
it's the default

File: plugins/callback/journald.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        description: unit name the logs are written against
```
Comment (Andersson007) [inline plugins/callback/journald.py:20]:
```suggestion
        description: Unit name the logs are written against.
```

File: plugins/callback/journald.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        env:
+          - name: JOURNALD_LOGGER_NAME
+        ini:
+          - section: callback_journald
+            key: logger_name
+      logger_level:
+        required: False
```
Comment (Andersson007) [inline plugins/callback/journald.py:27]:
```suggestion
```

File: plugins/callback/journald.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        description:
+        - Logging levels
+        - Valid values are 'notset', 'debug', 'info', 'warning', 'error', and 'critical'
+        - Default value is set to 'info'
```
Comment (Andersson007) [inline plugins/callback/journald.py:31]:
```suggestion
        - Logging levels.
        - Valid values are 'notset', 'debug', 'info', 'warning', 'error', and 'critical'.
        - Default value is set to 'info'.
```

Comment (Andersson007) [reply plugins/callback/journald.py:31]: i thinks we need to use `type:`, `default:`, etc. here and everywere

**Comment by Andersson007** [comment]
we also need CI tests. You can find the info about them in the ansible doc and examples in the source tree in `tests/integration/targets` (maybe in github.com/ansible/ansible repository as well)

File: plugins/callback/journald.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    def set_options(self, task_keys=None, var_options=None, direct=None):
+        super(CallbackModule, self).set_options(task_keys=task_keys,
+                                                var_options=var_options,
+                                                direct=direct)
+
+        if HAS_SYSTEMD:
+            self.logger_name = self.get_option('logger_name') or 'ansible'
+            self.logger = logging.getLogger(self.logger_name)
+            self.logger.addHandler(JournalHandler())
+            self.logger.setLevel(self.get_logger_level(self.get_option('logger_level').lower()))
```
Comment (vgavro) [inline plugins/callback/journald.py:94]: NoneType has no attribute lower (if log_level not specified - ansible 2.9.5)

File: plugins/callback/journald.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            self.logger_method = self.get_logger_method(self.get_option)
```
Comment (vgavro) [inline plugins/callback/journald.py:95]: self has no attribute "get_logger_method" (ansible 2.9.5)

**Comment by Andersson007** [comment]
@chestm007 any updates on this? This should be fixed (while reviews provided) or closed (if it's not going to be completed).
needs_info

**Comment by vgavro** [comment]
@Akasurde @Andersson007 even after fixes I was not satisfied by this plugin because of lack of information that was sent in journald, so i created a new one (based on logstash plugin) https://github.com/ansible-collections/community.general/pull/964 - I think we should use it instead of current pull request, if it's ok for you.

**Comment by Andersson007** [comment]
anyway, there's no feedback from the original author, so i'll close the PR
@vgavro thanks, will take a look later

**Pull request closed without merging**